File: app/src/main/java/com/dm/bomber/services/FormService.java
package com.dm.bomber.services;

import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;

public abstract class FormService extends Service {

    protected String url;
    protected String method;

    protected Request.Builder request;
    protected FormBody.Builder builder;

    public FormService(String url, String method, int... countryCodes) {
        super(countryCodes);

        this.url = url;
        this.method = method;
    }

    public FormService(String url, int... countryCodes) {
        this(url, "POST", countryCodes);
    }

    public void run(OkHttpClient client, Callback callback, Phone phone) {
        request = new Request.Builder();
        builder = new FormBody.Builder();

        buildBody(phone);

        request.url(url);
        request.method(method, builder.build());

        client.newCall(request.build()).enqueue(callback);
    }

    public abstract void buildBody(Phone phone);
}

File: app/src/main/java/com/dm/bomber/services/core/FormService.kt
package com.dm.bomber.services.core

import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request

abstract class FormService(
    protected val url: String,
    protected val method: String,
    vararg countryCodes: Int
) : Service(*countryCodes) {

    protected var request: Request.Builder = Request.Builder()

    protected lateinit var builder: FormBody.Builder

    constructor(url: String, vararg countryCodes: Int) : this(url, "POST", *countryCodes)

    override fun run(client: OkHttpClient, callback: Callback, phone: Phone) {
        builder = FormBody.Builder()

        buildBody(phone)

        request.url(url)
        request.method(method, builder.build())

        client.newCall(request.build()).enqueue(callback)
    }

    abstract fun buildBody(phone: Phone)
}
